### PR TITLE
GR716: Implement plo copy to SRAM

### PR DESCRIPTION
**Pull request by lukileczo**
## Motivation and Context

JIRA: RTOS-450
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `sparcv8leon3-gr716-mini`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/sparcv8leon3/_init.S [hunk 1]
```diff
@@ -18,11 +18,118 @@
...
+.extern _plo_load_addr
+.extern _plo_size
+
 .section ".init", "ax"
 .align 4
 .global _init
 .type _init, #function
 _init:
+	/* get current PC */
+	call . + 8
```
Comment (agkaminski) [inline hal/sparcv8leon3/_init.S:30]: Won't this cause us to jump to SRAM (regardless if it was initialized and if plo has been copied there)? `.` is provided by the compiler and it thinks we're executing from SRAM.

File: hal/sparcv8leon3/_traps.S [hunk 2]
```diff
@@ -43,9 +43,9 @@
 
 #define RESET \
 	mov	%g0, %g4; \
-	sethi %hi(_init), %g4; \
-	jmp %g4 + %lo(_init); \
-	nop;
+	nop; \
+	call _init; \
```
Comment (agkaminski) [inline hal/sparcv8leon3/_traps.S:47]: Why replace with call?

Side note - this is somewhat riscy strategy, same HW can be in an unexpected state after this kind of warm reset. Is there's no way to generate MCU reset in software?

File: hal/sparcv8leon3/_traps.S [hunk 3]
```diff
@@ -43,9 +43,9 @@
...
+	nop; \
```
Comment (agkaminski) [inline hal/sparcv8leon3/_traps.S:48]:
```suggestion
	nop;
```

File: hal/sparcv8leon3/gr716/gr716.c [hunk 4]
```diff
@@ -139,7 +139,7 @@ static void _gr716_pllSetDefault(void)
 
 
 /* Initialize external SRAM */
-static void _gr716_sram_init(void)
+void _gr716_sram_init(void)
```
Comment (agkaminski) [inline hal/sparcv8leon3/gr716/gr716.c:142]: I guess it became a dead code now, init is done in asm

File: hal/sparcv8leon3/_traps.S (hunk 2 above)
Comment (lukileczo) [reply hal/sparcv8leon3/_traps.S:47]: call is pc relative, so this allows for correct jump even when we're loaded to incorrect address

Comment (agkaminski) [reply hal/sparcv8leon3/_traps.S]: It is indeed, you're right

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**